metadata-ingestion/src/datahub/ingestion/source/kafka.py

```diff
@@ -265,10 +304,71 @@ def _extract_record(self, topic: str) -> Iterable[MetadataWorkUnit]:
                 self.report.report_workunit(wu)
                 yield wu
 
+    def build_custom_properties(self, topic_detail, extra_topic_config):
+
+        replication_factor = None
+        for _, p_meta in topic_detail.partitions.items():
+            if replication_factor is None or len(p_meta.replicas) > replication_factor:
+                replication_factor = len(p_meta.replicas)
+
+        custom_props = {
+            "Partitions": len(topic_detail.partitions),
+            "Replication Factor": replication_factor,
+        }
+        if extra_topic_config is not None:
+            for config_key in KafkaTopicConfigKeys:
+                custom_props[config_key] = self._get_config_value_if_present(
+                    extra_topic_config, config_key
+                )
+
+        return {
+            k: (json.dumps(v) if not isinstance(v, str) else v)
+            for k, v in custom_props.items()
+            if v is not None
+        }
+
     def get_report(self) -> KafkaSourceReport:
         return self.report
 
     def close(self) -> None:
         if self.consumer:
             self.consumer.close()
         super().close()
+
+    def _get_config_value_if_present(
+        self, config_dict: Dict[str, confluent_kafka.admin.ConfigEntry], key: str
+    ) -> Any:
+        return config_dict[key].value if key in config_dict.keys() else None
+
+    def fetch_extra_topic_details(self, topics: List[str]) -> Dict[str, dict]:
+        logger.info("Fetching config details for all topics")
+        extra_topic_details = {}
+        configs: Dict[
+            confluent_kafka.admin.ConfigResource, concurrent.futures.Future
+        ] = self.admin_client.describe_configs(
+            resources=[
+                confluent_kafka.admin.ConfigResource(
+                    confluent_kafka.admin.ResourceType.TOPIC, t
+                )
+                for t in topics
+            ]
+        )
+        logger.debug("Waiting for config details futures to complete")
+        concurrent.futures.wait(configs.values())
+        logger.debug("Config details futures completed")
+        for config_resource, config_result_future in configs.items():
+            try:
+                assert config_result_future.done()
+                assert config_result_future.exception() is None
```

Comment:
Do these asserts get caught by the Except? Shouldn't exit the program

Reply:
yes, they do get caught by surrounding try..except